components/raftstore-v2/tests/integrations/test_split.rs

```diff
@@ -0,0 +1,209 @@
+// Copyright 2022 TiKV Project Authors. Licensed under Apache-2.0.
+
+use std::{thread, time::Duration};
+
+use futures::executor::block_on;
+use kvproto::{
+    metapb, pdpb,
+    raft_cmdpb::{
+        AdminCmdType, AdminRequest, CmdType, RaftCmdRequest, RaftCmdResponse, Request, SplitRequest,
+    },
+};
+use raftstore_v2::router::PeerMsg;
+use tikv_util::store::new_peer;
+
+use crate::cluster::{Cluster, TestRouter};
+
+fn new_batch_split_region_request(
+    split_keys: Vec<Vec<u8>>,
+    ids: Vec<pdpb::SplitId>,
+    right_derive: bool,
+) -> AdminRequest {
+    let mut req = AdminRequest::default();
+    req.set_cmd_type(AdminCmdType::BatchSplit);
+    req.mut_splits().set_right_derive(right_derive);
+    let mut requests = Vec::with_capacity(ids.len());
+    for (mut id, key) in ids.into_iter().zip(split_keys) {
+        let mut split = SplitRequest::default();
+        split.set_split_key(key);
+        split.set_new_region_id(id.get_new_region_id());
+        split.set_new_peer_ids(id.take_new_peer_ids());
+        requests.push(split);
+    }
+    req.mut_splits().set_requests(requests.into());
+    req
+}
+
+fn must_split(region_id: u64, req: RaftCmdRequest, router: &mut TestRouter) {
+    let (msg, sub) = PeerMsg::raft_command(req);
+    router.send(region_id, msg).unwrap();
+    block_on(sub.result()).unwrap();
+
+    // todo: when persistent implementation is ready, we can use tablet index of
+    // the parent to check whether the split is done. Now, just sleep a second.
+    thread::sleep(Duration::from_secs(1));
+}
+
+fn put(
+    store_id: u64,
+    router: &mut TestRouter,
+    region: &metapb::Region,
+    key: &[u8],
+) -> RaftCmdResponse {
+    let mut req = RaftCmdRequest::default();
+    req.mut_header().set_region_id(region.id);
+    req.mut_header()
+        .set_region_epoch(region.get_region_epoch().clone());
+    req.mut_header().set_peer(
+        region
+            .get_peers()
+            .iter()
+            .find(|p| p.get_store_id() == store_id)
+            .unwrap()
+            .clone(),
+    );
+
+    let mut put_req = Request::default();
+    put_req.set_cmd_type(CmdType::Put);
+    put_req.mut_put().set_key(key.to_vec());
+    put_req.mut_put().set_value(b"v1".to_vec());
+    req.mut_requests().push(put_req);
+
+    let (msg, mut sub) = PeerMsg::raft_command(req.clone());
+    router.send(region.id, msg).unwrap();
+    assert!(block_on(sub.wait_proposed()));
+    assert!(block_on(sub.wait_committed()));
+    block_on(sub.result()).unwrap()
+}
+
+fn must_put_succeed(store_id: u64, router: &mut TestRouter, region: &metapb::Region, key: &[u8]) {
+    let resp = put(store_id, router, region, key);
+    assert!(!resp.get_header().has_error(), "{:?}", resp);
+}
+
+fn must_put_error(store_id: u64, router: &mut TestRouter, region: &metapb::Region, key: &[u8]) {
```

Comment:
Better implement these inline, it's hard to read to jump back and forth.

Reply:
Done